src/ansi/mod.rs

```diff
@@ -51,20 +46,60 @@ pub fn truncate_str<'a>(s: &'a str, display_width: usize, tail: &str) -> Cow<'a,
-                    result.push_str(&" ".repeat(display_width.saturating_sub(used)));
+                let width_of_grapheme = g.width();
+                if used + width_of_grapheme > display_width {
+                    // Handle case "2." mentioned in `truncate_str` docs and fill the
+                    // hole left by double-width (2w) truncation.
+                    if let Some(fillchar) = fill2w {
+                        if width_of_grapheme == 2 && used < display_width {
+                            result.push(fillchar);
+                        } else if width_of_grapheme > 2 {
+                            // Should not happen, this means either unicode_segmentation
+                            // graphemes are too wide, or the unicode_width is calculated wrong.
+                            // Fallback:
+                            debug_assert!(width_of_grapheme <= 2, "strange grapheme width");
+                            for _ in 0..display_width.saturating_sub(used) {
+                                result.push(fillchar);
+                            }
+                        }
+                    }
                     break;
                 }
                 result.push_str(g);
-                used += w;
+                used += width_of_grapheme;
             }
         } else {
             result.push_str(t);
         }
     }
 
-    Cow::from(format!("{result}{result_tail}"))
+    result.push_str(&result_tail);
+    Cow::from(result)
+}
+
+/// Truncate string such that `tail` is present as a suffix, preceded by as much of `s` as can be
+/// displayed in the requested width. Even with `tail` empty the result may not be a prefix of `s`.
+// Return string constructed as follows:
+// 1. `display_width` characters are available. If the string fits, return it.
+//
+// 2. If a double-width (fullwidth) grapheme has to be cut in the following steps, replace the first
+//    half with a space (' '). If this happens the result is no longer a substring of the input!
+//
+// 3. Contribute graphemes and ANSI escape sequences from `tail` until either (1) `tail` is
+//    exhausted, or (2) the display width of the result would exceed `display_width`.
+//
+// 4. If tail was exhausted, then contribute graphemes and ANSI escape sequences from `s` until the
+//    display_width of the result would exceed `display_width`.
+pub fn truncate_str<'a>(s: &'a str, display_width: usize, tail: &str) -> Cow<'a, str> {
+    truncate_str_impl(s, display_width, tail, Some(' '))
+}
+
+/// Truncate string `s` so it fits into `display_width`, ignoring any ANSI escape sequences when
+/// calculating the width. If a double-width ("fullwidth") grapheme has to be cut, it is omitted and
+/// the resulting string is *shorter* than `display_width`. But this way the result is always a
+/// prefix of the input `s`.
+pub fn truncate_str_short(s: &str, display_width: usize) -> Cow<str> {
+    truncate_str_impl(s, display_width, "", None)
```

Comment:
clippy wants `#[allow(dead_code)]` here

Reply:
I'll just merge it with the  --max-syntax-highlighting-length PR :)